### PR TITLE
Add immune deconvolution module

### PR DESCRIPTION
**Pull request by komalsrathi**
#### What scientific question is your analysis addressing?

Deconvolution of immune cell types from bulk RNA-seq data.

#### What was your approach?

The goal of this analysis is to use the R package `immunedeconv` to quantify and compare various immune cell types in the tumor microenvironment (TME) across various cancer and GTEx groups. 

The package `immunedeconv`, provides the following deconvolution methods: `xCell` (n = 64; immune and non-immune cell types), `CIBERSORT` (relative mode; n = 22 immune cell types); `CIBERSORT (abs.)` (absolute mode; n = 22 immune cell types), `TIMER` (n = 6), `EPIC` (n = 6), `quanTIseq` (n = 10) and `MCP-Counter` (n = 8). 

Based on literature (PMID: 31641033; PMID: 31510660) on benchmarking various methods, `xCell` and `CIBERSORT` are recommended and widely used. `xCell` and `CIBERSORT (abs.)` output enrichment scores whereas `CIBERSORT` outputs actual cell fractions so we will use `xCell` and `CIBERSORT (abs.)` to enable comparisons across various samples or groups. 

#### What GitHub issue does your pull request address?

https://github.com/PediatricOpenTargets/ticket-tracker/issues/217

#### Which areas should receive a particularly close look?

1. I have generated the outputs for both `xCell` and `CIBERSORT (abs.)` by running the command `RUN_CIBERSORT=1 bash run-immune-deconv.sh`. Because CIBERSORT needs two files (details in the README) that are not to be publicly pushed, we cannot run CIBERSORT via CI. So, in order to test with CI, we can use the command `bash run-immune-deconv.sh` which will only run `xCell`. We can either keep CIBERSORT or remove it depending on feedback.

2. `xCell` depends on the variability of samples in a group to output transformed scores. So, my approach was to split the full expression matrix into individual groups of `cohort + cancer_group` in case of Tumor samples or `cohort + gtex_group` in case of GTEx samples and process each group separately. Once processed, I am combining all the output into a single `rds` file which can be used for plotting. 

#### Is there anything that you want to discuss further?

1. Feedback on plotting 
2. The module takes a while to run. For `xCell`, the runtime is about 1 hour on my local machine but took 5 hours 30 mins when running via docker and for `CIBERSORT (abs.)`, it is about 3 hours on my local machine but took 4 hours 25 mins via docker. So if someone else could also test this, it would be great. The maximum time spent is for the GTEx tissues because of the large sample size. If we don't run this module for GTEx normal tissues, it will reduce the runtime significantly. 

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?

Yes

#### What types of results are included (e.g., table, figure)?

rds files with immune scores
heatmaps of average immune scores per group (cohort + cancer_group or cohort + gtex_group)

#### What is your summary of the results?

N/A

#### Reproducibility Checklist

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

#### Documentation Checklist

- [x] This analysis module has a `README` and it is up to date.
- [ ] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

### CODE REVIEW

**Comment by komalsrathi** [comment]
@runjin326 please feel free to add other reviewers (and remove yourself if not relevant to you).

**Review by runjin326 — COMMENTED** [review]
@komalsrathi - thanks so much for preparing this! I ran the script only with xCell in docker and also it took me 5hr 15 minutes to finish running. The results are as expected - and I like the heatmap. 
One thing I noticed is that when I run the 02 script for the first time, it says `rlist` is not available. We might need to add that to the docker file as well.
I am currently running cibersort and if that finishes without issue, I would approve this with the addition of rlist to docker.

**Comment by komalsrathi** [comment]
@runjin326 I actually added `rlist` to the Dockerfile here: https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/immune-deconv/Dockerfile#L213-L215. I am not sure why it is not be able to find the package.

**Comment by runjin326** [comment]
> @runjin326 I actually added `rlist` to the Dockerfile here: https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/immune-deconv/Dockerfile#L213-L215. I am not sure why it is not be able to find the package.

Oh my bad - missed that change. The Docker change would not be able to change until we merge it and then pull the docker again so that is why it is not installed yet. Will be available after we merge this. 

I completed the `CIBERSORT (abs.)` run and everything works fine.

**Review by runjin326 — APPROVED** [review]
@komalsrathi , thanks for working on this! Everything looks good to me. 
Could you please do another dev merge to this branch? I will merge right after - assuming the docker check is fine. It is currently failed.

**Comment by komalsrathi** [comment]
I am not sure why the docker is [failing](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/runs/4748325700?check_suite_focus=true#step:5:5173) - it looks unrelated to this PR. Any suggestions how we can fix this?

**Comment by runjin326** [comment]
I just hit the re-run for the check jobs - hopefully this fixes it just like last time. Does this branch have the latest dev?

**Comment by komalsrathi** [comment]
Just pulled in the latest dev - Hope it works.

**Comment by runjin326** [comment]
@komalsrathi looks like the error message is coming from the docker build for `immunedeconv` - just wondering: is there another way to install this package? If putting it in docker generates error, would it be possible to just install in the module with `install.packages()`? @logstar - some input here?

**Comment by komalsrathi** [comment]
I don't think immunedeconv is on CRAN or Bioconductor so that's the way I have been installing it per their [README](https://github.com/icbi-lab/immunedeconv#standard-r-package). It worked for [OpenPBTA](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/master/Dockerfile#L211-L212) and for OT, I am just installing a more updated version, but this is a valid command: `remotes::install_github('icbi-lab/immunedeconv', ref = 'c70539f2b08901687561dca755337fc6a5130440', dependencies = TRUE)` so ideally it should work.

**Comment by logstar** [comment]
Maybe add `RUN ./install_bioc.r XML` before the `RUN R -e "remotes::install_github('icbi-lab/immunedeconv', ref = 'c70539f2b08901687561dca755337fc6a5130440', dependencies = TRUE)"` command would work. The docker compile failed on loading the R XML package.

Screen Shot 2022-01-19 at 12 24 36 PM

**Comment by logstar** [comment]
The docker build failed with the same error after adding `RUN ./install_bioc.r XML`.

I was wondering whether the older version of immunedeconv `493bcaa9e1f73554ac2d25aff6e6a7925b0ea7a6` could still be used here. This version is used in OpenPBTA. The new version `c70539f2b08901687561dca755337fc6a5130440` requires XML installation from <https://github.com/omegahat/XML>, which caused the XML package compilation error. I am not sure how to fix the XML compilation error. I guess it is caused by incompatible libxml2 version.

The older version of immunedeconv could probably be installed without any error.

**Comment by runjin326** [comment]
@logstar , thanks for looking into this - maybe we can just use the older version here so avoid the compatibility issue? (Assuming the older version can still be run here without causing result differences etc.)

**Comment by komalsrathi** [comment]
The older version does not have some of the functions so I will have to make some changes. I will update and let you know.

File: Dockerfile
Comment (logstar) [inline Dockerfile]:
```suggestion
RUN R -e "remotes::install_github('dviraran/xCell')"
RUN R -e "remotes::install_github('GfellerLab/EPIC')"
RUN R -e "remotes::install_github('ebecht/MCPcounter/Source')"

RUN ./install_bioc.r \
    data.tree \
    limSolve \
    DiagrammeR \
    quadprog \
    lpSolve \
    influenceR \
    rgexf \
    visNetwork \
    Rook

RUN R -e "remotes::install_github('icbi-lab/immunedeconv', ref = 'c70539f2b08901687561dca755337fc6a5130440', dependencies = FALSE)"
```

Comment (komalsrathi) [reply Dockerfile]: @logstar let's keep this as a second option. I converted this PR to draft as I am testing the old version and if there is no diff in the output, then we should be good.

**Comment by komalsrathi** [comment]
@runjin326 @logstar I downgraded to the older package, luckily the functions are compatible between the old and new versions and no change had to made to the code. I double checked the old and newly generated rds files as well as the heatmaps and they match so we can go ahead with the old package version. thanks both!

**Comment by runjin326** [comment]
Thanks so much @komalsrathi! This looks good to me - maybe @logstar can also review this? Thanks!

**Review by logstar — APPROVED** [review]
Thank you for fixing the docker build error @komalsrathi .

I reviewed the latest Dockerfile and docker build action. They both look good to me.

**Pull request merged**
